src/dlsproto/client/request/internal/GetRange.d

```diff
@@ -317,422 +160,631 @@ public struct GetRange
+                    {
+                        payload.addConstant(MessageType.Ack);
+                    }
+            );
 
-        return ReceivedMessageAction.Continue;
+            this.record_stream.stop();
+        }
     }
 
-    /***************************************************************************
-
-        Debug message, printed on state change.
 
-    ***************************************************************************/
+    /**************************************************************************
 
-    debug (ClientGetRangeState):
+        The fiber that handles user controlled signals.
 
-    import ocean.io.Stdout;
+    **************************************************************************/
 
-    private void beforeState ( )
+    private class Controller
     {
-        static char[][] machine_msg =
-        [
-            State.EnsuringConnEstablished: "EnsuringConnEstablished",
-            State.Initialising: "Initialising",
-            State.Receiving: "Receiving",
-            State.RequestingStateChange: "RequestingStateChange",
-            State.Exit: "Exit",
-        ];
-
-        alias typeof(this.context.shared_working.desired_state) DesiredState;
-
-        static char[][] request_msg =
-        [
-            DesiredState.None: "???",
-            DesiredState.Running: "Running",
-            DesiredState.Suspended: "Suspended",
-            DesiredState.Stopped: "Stopped"
-        ];
-
-        Stdout.green.formatln("GetRange state: Machine = {}, Request = {}",
-            machine_msg[this.state],
-            request_msg[this.context.shared_working.desired_state]).default_colour;
+        /// The fiber.
+        private MessageFiber fiber;
+
+        /// The `RecordStream` to notify when the request is resumed.
+        private RecordStream record_stream;
+
+        /**********************************************************************
+
+            Constructor, starts the fiber.
+
+            Params:
+                record_stream = the `RecordStream` to notify when the request is
+                resumed.
+
+        **********************************************************************/
+
+        private this (RecordStream record_stream)
+        {
+            this.record_stream = record_stream;
+            this.fiber = this.outer.resources.getFiber(&this.fiberMethod);
+            this.fiber.start();
+        }
+
+        /**********************************************************************
+
+            Waits for controller signals and handles them. Terminates on
+            `FiberSignal.StopController`
+
+        **********************************************************************/
+
+        private void fiberMethod ()
+        {
+            while (true)
+            {
+                auto event = this.outer.request_event_dispatcher.nextEvent(
+                    this.fiber,
+                    Signal(ControllerSignal.Resume),
```

Comment:
What handles `controller.suspend`?

Reply:
Nothing, we're not expecting this signal asynchronously, but we're checking it before passing each record to the user:

https://github.com/sociomantic-tsunami/dlsproto/pull/39/files#diff-c6938f346e8e72b5984b4ebffc36d024R505